### ISSUE TITLE
tf.contrib.data.parallel_interleave with tf.data.Dataset.from_generator

### ISSUE THREAD

**Issue by dmus** [tensorflow/tensorflow#19269]
I think this is a feature request (or maybe I miss something). When data is coming from multiple remote sources multiple tf.data.Dataset.from_generator should run in parallel. Therefore tf.contrib.data.parallel_interleave is already available, see this small code example.

```
def generator(n):
  # returns n-th generator function

def dataset(n):
  return tf.data.Dataset.from_generator(generator(n))

ds = tf.data.Dataset.range(N).apply(tf.contrib.data.parallel_interleave(dataset, cycle_lenght=N))

```
# where N is the number of generators you use

### Describe the problem
This problem with the code above is that generator cannot look like this:
def generator(n):
  return consumers[n] # A python list with n items

Because then the error 'TypeError: list indices must be integers or slices, not Tensor' will occur. However it would be really nice to have support for multiple from_generator pipelines in parallel.

### Source code / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Comment by mrry** [comment]
It's possible with some involved usage of [`tf.py_func()`](https://www.tensorflow.org/api_docs/python/tf/py_func), but we've had the feature requested enough times that it's worth adding to the `Dataset.from_generator()` API. An implementation is currently out for review, and it will add an optional `args` argument to `Dataset.from_generator()`, allowing you to specify tensors that should be passed to the generator function as NumPy arrays (like the `inp` argument to `tf.py_func()`).

/cc @jsimsa FYI.

**Comment by mrry** [comment]
The implementation has been submitted and will become available in the next sync. To work around the problem using `tf.py_func()` in a `Dataset.map()`, you'd have to maintain a list of Python iterator objects, then write a Python function that takes an integer, looks up the corresponding Python iterator in the list, and then returns the result of calling `next()` on it. (This is essentially all that `Dataset.from_generator()` does, although there's some additional nuance to ensuring that the iterators are properly disposed of at the end of the computation.)
